src/WebJobs.Script.WebHost/Management/LinuxInstanceManager.cs

```diff
@@ -44,70 +37,102 @@ public LinuxInstanceManager(IHttpClientFactory httpClientFactory, IScriptWebHost
 
         public abstract Task<string> SpecializeMSISidecar(HostAssignmentContext context);
 
-        public bool StartAssignment(HostAssignmentContext context)
+        public async Task<bool> AssignInstanceAsync(HostAssignmentContext context)
         {
-            if (!_webHostEnvironment.InStandbyMode)
+            if (!IsValidEnvironment(context))
             {
-                // This is only true when specializing pinned containers.
-                if (!context.Environment.TryGetValue(EnvironmentSettingNames.ContainerStartContext, out string startContext))
+                return false;
+            }
+
+            if (context.IsWarmupRequest)
+            {
+                await HandleWarmupRequestAsync(context);
+                return true;
+            }
+
+            lock (_assignmentLock)
+            {
+                if (_assignmentContext == null)
+                {
+                    _assignmentContext = context;
+                    _assignment = AssignAsync(context);
+                }
+                else if (!_assignmentContext.Equals(context))
                 {
-                    _logger.LogError("Assign called while host is not in placeholder mode and start context is not present.");
                     return false;
                 }
             }
 
-            if (_environment.IsContainerReady())
+            await _assignment;
+            return true;
+        }
+
+        public bool StartAssignment(HostAssignmentContext context)
+        {
+            if (!IsValidEnvironment(context))
             {
-                _logger.LogError("Assign called while container is marked as specialized.");
                 return false;
             }
 
             if (context.IsWarmupRequest)
             {
-                // Based on profiling download code jit-ing holds up cold start.
-                // Pre-jit to avoid paying the cost later.
-                Task.Run(async () => await DownloadWarmupAsync(context.GetRunFromPkgContext()));
+                Task.Run(async () => await HandleWarmupRequestAsync(context));
                 return true;
             }
-            else if (_assignmentContext == null)
+
+            lock (_assignmentLock)
             {
-                lock (_assignmentLock)
+                if (_assignmentContext != null)
                 {
-                    if (_assignmentContext != null)
-                    {
-                        return _assignmentContext.Equals(context);
-                    }
-                    _assignmentContext = context;
+                    return _assignmentContext.Equals(context);
                 }
+                _assignmentContext = context;
+            }
 
-                _logger.LogInformation($"Starting Assignment. Cloud Name: {_environment.GetCloudName()}");
+            _ = AssignAsync(context);
```

Comment:
We should this call into the `lock` scope and assign it to `_assignment`.

No need to await in this method, but doing that will ensure the `Task` is assigned and subsequent calls to `AssignInstanceAsync` would await on the same `Task`. Skipping this might create a condition where you might run into a null reference in line 66